Multiple upload

Elfangor93 opened the pull request:
Multi-upload with [uppy](https://github.com/transloadit/uppy), [tus](https://tus.io/) and [async-sema](https://github.com/vercel/async-sema).

This PR adds a form to upload multiple images into the same category.
You find the form in the Joomla backend in Components > JoomGallery > Images > New > Multiple New.

All configuration settings which are not red in the tab "General Settings" (level one tab) are taken into account and will influence the result or the upload process itself.

## Upload form

![grafik](https://user-images.githubusercontent.com/39154009/236674358-f8799f49-a5bb-46b1-88b5-2d46be2f6eef.png)

On the right hand side the general image data can be inserted, the same way is it was done in JG3 Drag'N'Drop uploader. Data enterd here is used for all the images.
Inside the uppy form to the left there is a edit button below each image preview allowing to adjust title, author and description for each image individually. This data on each image will override the general data from the right form if available.
If activated in the configuration the image metadata will override the general and individual data from the upload form.
(Priority: Metadata > individual image data > general form data)

Below the uppy form the debug mode can be activated. This works similarly as in JG3 Drag'N'Drop uploader. But the uploader will display debug data additionally in the browser console.


## Request protocol for one image

![grafik](https://user-images.githubusercontent.com/39154009/236674291-a322be74-ead4-46f1-8351-8fa0c8cb39cf.png)
Multiple requests to the server are perfomed for each upload. Except for the last POST requests, all the requests are from the tus communication protocol uploading the files to the server.

## System architecture

- [Link to project presentation](https://docs.google.com/presentation/d/1kXGfGRrHswU0M3yh0zUvOwW1fYqB07cksHzNxzcEyxI/edit?pli=1#slide=id.g2402bc76c11_1_5)

## Test instructions

- Component has to be new installed. There are changes in the DB which need to be applied.
- Since this PR adds client side application (uppy form), we need to test on different browsers.
- PR was coded using PHP 7.4.x. Has to bested on PHP 8.x.
- If you find a bug please post as much as possible
     - Error message in the view
     - Browser console log (Browser developer tools -> Console tab)
     - Request response of the failing request (Browser developer tools -> Network analysis tab)
     - System details (PHP version, ...)

AlexanderSupp commented:
I tested the actual status. Multiple uploads works. Images stored in temp file. It's great to see how multiple images load in the same time.
I wait for next action to see the images in the normal JG4 file structure.
BTW:
I uninstalled JG4 and then I installed multiple-upload. This installation process does not work.
I installed JG4 main and then multiple-upload and that works.

reni68 commented:
> I uninstalled JG4 and then I installed multiple-upload. This installation process does not work.

Here too:
![94_1](https://user-images.githubusercontent.com/126355001/234368216-9af1c4ef-b5f6-41f9-801f-4d540922db48.jpeg)

reni68 commented:
But even after installing Main and then "Multiple Upload" I get this error when selecting "Multiple New" when uploading images:
![94_2](https://user-images.githubusercontent.com/126355001/234369381-0df133b9-a0cb-497a-ba26-3d67e8ca4dfd.jpeg)
I am waiting with more tests

MrMusic commented:
A hint from me:
Thanks for testing, but this PR status is still on 'Draft', which means it is still in progress and not yet finished...

AlexanderSupp commented:
I uninstalled JG4 and installed only multiple-upload.
I tried to upload a single image with new.
The result was:
![grafik](https://user-images.githubusercontent.com/110494397/236804954-9dbb27bf-acda-4124-bee3-f675b013f14c.png)
Now I try multiple upload.

eumel1602 commented:
I looked it up and found the following:
- if I don't assign a title for the picture(s) and/or if I don't assign a category, the upload doesn't work, but you don't see an error message stating the error or what's missing
- If I want to edit something using the pencil symbol before uploading an image, I can then save it. Unfortunately, it only saves the field that I edited last. So: I can change three things (title, description and author) if I change all three at once and then save, then all three must also be changed. Currently only the last thing I edited is saved.
- If I change the "title" with the pen symbol and save, this change is not visible in the image, only when the upload starts is it changed during the upload
- When multi-uploading, there is a field at the bottom right: "Description" when I write something in there, I couldn't find it later in the individual images. What's the point of the field?
- I could not upload .webP or .png images. The upload worked, but they are marked with a "yellow" circle in the upper right corner and are not displayed in the image manager...(the set debug mode cannot be called up, the link is not visible)

I tested with Win / Joomla 4.3.1 / php 8.0.20 / Browser: FF , Edge and Chrome. the errors are the same in all three browsers.

AlexanderSupp commented:
The same photo with multiple new.
**Info-File:**
{"id":"69678a4aff7219fee6e0bedeb4499328","size":721679,"offset":721679,"extension":"jpg","filename":"Blumenmeer - Kopie.jpg","mimetype":"image\/jpeg","ispartial":false,"isfinal":true,"relativepath":"null","name":"Blumenmeer - Kopie.jpg","type":"image\/jpeg","owner":"Alexanders Fotos","filetype":"image\/jpeg"}
**Checked JPG-File (in Temp with no extension) and this looks fine.**
The Owner is correct. I modified this as author, before upload.
But the Author is not the owner.

![grafik](https://user-images.githubusercontent.com/110494397/236808732-bd981399-83f8-4492-bddb-314a01451e6f.png)

I'm waiting for storing the results into the JG4-Files.

eumel1602 commented:
> I uninstalled JG4 and installed only multiple-upload. I tried to upload a single image with new. The result was: ![grafik](https://user-images.githubusercontent.com/110494397/236804954-9dbb27bf-acda-4124-bee3-f675b013f14c.png) Now I try multiple upload.

Yes, that's the case for me too, a single upload doesn't work. But it can be that this doesn't have to work in this PR?!?

Elfangor93 commented:
> I tried to upload a single image with new.
The result was:
Now I try multiple upload.

Confirmed and fixed. Upload a single image works again. Thank you for reporting!

Elfangor93 commented:
> If I want to edit something using the pencil symbol before uploading an image, I can then save it. Unfortunately, it only saves the field that I edited last. So: I can change three things (title, description and author) if I change all three at once and then save, then all three must also be changed. Currently only the last thing I edited is saved.

Confirmed. Since this is an issue of the uppy library this has to be fixed in the uppy project. Bug is already reported in the corresponding repository:
https://github.com/transloadit/uppy/issues/4427

Elfangor93 commented:
> If I change the "title" with the pen symbol and save, this change is not visible in the image, only when the upload starts is it changed during the upload

Thats a feature not a bug. The title of the image preview in the dashboard only changes after the image is successfully uploaded and saved. Before that the original filename is shown.
Do you think this should be changed? If yes, why?

eumel1602 commented:
> > If I change the "title" with the pen symbol and save, this change is not visible in the image, only when the upload starts is it changed during the upload
> 
> Thats a feature not a bug...  Do you think this should be changed? If yes, why?
My feeling on first use was why doesn't the editor rewrite that? I tried to change it again 2 times. After uploading the pictures for the first time, I knew it was possible. After that, I didn't think it mattered that much. I can't judge it technically, but if it works, it would be very nice for me.... (How much time does it take... is it related?!)

MrMusic commented:
In addition to the points already postet, I noticed the following:
- If 'Original filename' = Yes in the configuration in the 'Upload' tab, the upload does not work.

- When clicking on 'Upload' there is a Javascript error.

Elfangor93 commented:
>if I don't assign a title for the picture(s) and/or if I don't assign a category, the upload doesn't work, but you don't see an error message stating the error or what's missing

Confirmed and fixed.

Elfangor93 commented:
>When multi-uploading, there is a field at the bottom right: "Description" when I write something in there, I couldn't find it later in the individual images. What's the point of the field?

Confirmed. There was a bug in the code such that the description was not submitted with the form. Therefore no description text was saved. Should be solved now.

Elfangor93 commented:
>Info-File:
{"id":"69678a4aff7219fee6e0bedeb4499328","size":721679,"offset":721679,"extension":"jpg","filename":"Blumenmeer - Kopie.jpg","mimetype":"image/jpeg","ispartial":false,"isfinal":true,"relativepath":"null","name":"Blumenmeer - Kopie.jpg","type":"image/jpeg","owner":"Alexanders Fotos","filetype":"image/jpeg"}
Checked JPG-File (in Temp with no extension) and this looks fine.
The Owner is correct. I modified this as author, before upload.
But the Author is not the owner.

I used `owner` as the tus-matadata field-id of the Author field since the id `author` is already used by the uppy dashboard for something else. I renamed now all tus-matadata field-ids to avoid confusion. The new field-ids are:
- `jtitle` for the Title field
- `jdescription` for the Description field
- `jauthor` for the Author field.

Elfangor93 commented:
>if I add a picture and change the start numbering, it is ignored.

Confirmed and fixed.

Elfangor93 commented:
>I don't see the title of the first picture at all - only in the picture overview.

Can not confirm this.
This behavior looks very strange to me. It looks to me as if you are testing with an outdated code-base. Please deinstall and reinstall the component and see if the issue persists.
If the issue persists please give me a complete step by step guide how to reproduce the issue.

eumel1602 commented:
I've tried to double check @Elfangor93's changes and can confirm almost everything as fixed.
---------------------------------
- If I don't assign a category before uploading, I get the error message: "Please fill out the required fields." But you don't see "red" which field is missing...

I could not upload an image larger than 5 MB (error message: debug image)

![zugross](https://github.com/JoomGalleryfriends/JG4-dev/assets/96334985/2ef81563-5e45-4cc8-80f0-c367eb195a24)


---------------------------------
- If I e.g. 100 pictures via multiupload, I give each picture an extra description and each picture an extra author. But I want to give each picture a name. But I forgot to enter this under "Title" on the right and go to "Upload"! Then it gives me an error message BUT I can't get the images back. With the "retry" button or the "x" button, all images and all settings are gone!

![nicht zuzruck](https://github.com/JoomGalleryfriends/JG4-dev/assets/96334985/3a24c5d5-6639-4939-9096-742907f746bc)

Elfangor93 commented:
>I could not upload an image larger than 5 MB (error message: debug image)

@eumel1602 Does the same image work when you use single upload?

eumel1602 commented:
In the single upload it is included but not displayed.
Status report here:

eumel1602 commented:
Excuse me. It must be because of the picture. Other larger images work in single upload and also in multi-upload...

Elfangor93 commented:
>If I e.g. 100 pictures via multiupload, I give each picture an extra description and each picture an extra author. But I want to give each picture a name. But I forgot to enter this under "Title" on the right and go to "Upload"! Then it gives me an error message BUT I can't get the images back. With the "retry" button or the "x" button, all images and all settings are gone!

The form validation was in general very hacky developed. Before the form got validated when the uploads already started, thats why I had to cancel the uploads, but canceling uploads removed the files in the dashboard.
Now the form gets validated before the upload starts and the upload only get started when the form is correctly validated.

Thanks @eumel1602 for pointing me towards a much better solution for the same problem  ;-)

Elfangor93 commented:
>If I don't assign a category before uploading, I get the error message: "Please fill out the required fields." But you don't see "red" which field is missing...

There is no noce way to fix that. I tried something. Is that good enough?

AlexanderSupp commented:
I uploaded 40 files with Nmb parallel processes 1 and everything works.
I uploaded the same 40 files with Nmb parallel processes 10, and I get errors.
Some images are OK, other not. 
![Screenshot 2023-05-18 224028](https://github.com/JoomGalleryfriends/JG4-dev/assets/110494397/359f2dec-0373-476f-9adc-b89e1ad349e0)
![Screenshot 2023-05-18 223313](https://github.com/JoomGalleryfriends/JG4-dev/assets/110494397/1630c599-03af-4a97-aa6e-b62949a8029a)
![Screenshot 2023-05-18 223204](https://github.com/JoomGalleryfriends/JG4-dev/assets/110494397/fc6d19e0-7ad3-4f5d-b9e2-81b30b18807b)

AlexanderSupp commented:
More information about this:
<html><body>
<!--StartFragment--><br>
<font size="1">

( ! )  Notice: getimagesize(): Error reading from  C:\wamp64\www\Joomla\tmp\tmp_img.jpg! in  C:\wamp64\www\Joomla\administrator\components\com_joomgallery\src\Service\IMGtools\IMGtools.php  on line 495
--


1 | 0.0001 | 368184 | {main}(  ) | ...\index.php:0
2 | 0.0002 | 368952 | require_once( 'C:\wamp64\www\Joomla\administrator\includes\app.php ) | ...\index.php:32
3 | 0.0083 | 1386656 | Joomla\CMS\Application\CMSApplication->execute(  ) | ...\app.php:61
4 | 0.0086 | 1507872 | Joomla\CMS\Application\AdministratorApplication->doExecute(  ) | ...\CMSApplication.php:293
5 | 0.0477 | 2596784 | Joomla\CMS\Application\AdministratorApplication->dispatch( $component = ??? ) | ...\AdministratorApplication.php:186
6 | 0.0512 | 2870760 | Joomla\CMS\Component\ComponentHelper::renderComponent( $option = 'com_joomgallery', $params = ??? ) | ...\AdministratorApplication.php:143
7 | 0.0527 | 3075032 | Joomla\CMS\Dispatcher\ComponentDispatcher->dispatch(  ) | ...\ComponentHelper.php:361
8 | 0.0530 | 3096696 | Joomgallery\Component\Joomgallery\Administrator\Controller\JoomFormController->execute( $task = 'ajaxsave' ) | ...\ComponentDispatcher.php:143
9 | 0.0530 | 3096760 | Joomla\CMS\MVC\Controller\BaseController->execute( $task = 'ajaxsave' ) | ...\JoomFormController.php:87
10 | 0.0530 | 3096760 | Joomgallery\Component\Joomgallery\Administrator\Controller\ImageController->ajaxsave(  ) | ...\BaseController.php:678
11 | 0.0530 | 3096760 | Joomla\CMS\MVC\Controller\FormController->save( $key = ???, $urlVar = ??? ) | ...\ImageController.php:107
12 | 0.0669 | 3516520 | Joomgallery\Component\Joomgallery\Administrator\Model\ImageModel->save( $data = ['id'  => '0', 'uploader' => 'tus', 'multiple' => '1', 'uuid' =>  '5d140b029646c57d5a683cbb6bda9b06', 'imgtitle' => 'title',  'filecounter' => '2', 'catid' => '7', 'published' => '1',  'imgauthor' => '', 'language' => '*', 'debug' => '0', 'imgtext'  => '', 'access' => '1', 'tags' => []] ) | ...\FormController.php:675
13 | 0.3724 | 4159968 | Joomgallery\Component\Joomgallery\Administrator\Service\Uploader\Uploader->createImage( $data_row = class  Joomgallery\Component\Joomgallery\Administrator\Table\ImageTable {  protected $_tbl = '#__joomgallery'; protected $_tbl_key = 'id';  protected $_tbl_keys = [0 => 'id']; protected $_db = class  Joomla\Database\Mysql\MysqlDriver { public $name = 'mysql'; protected  $nameQuote = '`'; protected $nullDate = '0000-00-00 00:00:00'; protected  $utf8mb4 = TRUE; protected $mariadb = FALSE; protected $connection =  class PDO { ... }; private ${Joomla\Database\DatabaseDriver}database =  'JoomlaTest'; public $serverType = 'mysql'; protected $count = 88;  protected $cursor = NULL; protected $executed = FALSE; protected $limit =  0; protected $offset = 0; protected $options = [...]; protected $sql =  class Joomla\Database\Mysql\MysqlQuery { ... }; protected $statement =  class Joomla\Database\Pdo\PdoStatement { ... }; protected $tablePrefix =  'aji0d_'; protected $utf = TRUE; protected $errorNum = 0; protected  $errorMsg = NULL; protected $transactionDepth = 0; protected $factory =  class Joomla\Database\DatabaseFactory { ... }; protected $monitor =  NULL; private ${Joomla\Database\DatabaseDriver}dispatcher = class  Joomla\Event\Dispatcher { ... }; public $charset = 'utf8mb4' };  protected $_trackAssets = TRUE; protected $_rules = NULL; protected  $_locked = FALSE; protected $_autoincrement = TRUE; protected  $_columnAlias = ['published' => 'published']; protected $_jsonEncode =  []; protected $_supportNullValue = FALSE; public $typeAlias =  'com_joomgallery.image'; protected $_errors = []; private  ${Joomla\CMS\Table\Table}dispatcher = class Joomla\Event\Dispatcher {  protected $events = [...]; protected $listeners = [...] }; public $form =  class Joomla\CMS\Form\Form { protected $data = class  Joomla\Registry\Registry { ... }; protected $errors = [...]; protected  $name = 'image'; protected $options = [...]; protected $xml = class  SimpleXMLElement { ... }; public $repeat = FALSE; private  $databaseAwareTraitDatabase = NULL }; public $id = '164'; public $catid =  '7'; public $alias = 'as-sam-9288-jpg-2'; public $imgtitle =  'AS-SAM_9288.JPG'; public $imgtext = ''; public $imgauthor = ''; public  $imgdate = '2016:11:26 17:16:57'; public $imgmetadata =  '{"exif":{"IFD0":{"ImageDescription":"SAMSUNG  CSC","Make":"SAMSUNG","Model":"NX300","Orientation":1,"XResolution":"350\\/1","YResolution":"350\\/1","ResolutionUnit":2,"Software":"1.45  ","DateTime":"2016:11:26  17:16:57","YCbCrPositioning":2,"Exif_IFD_Pointer":254},"EXIF":{"ExposureTime":"1\\/40","FNumber":"35\\/10","ExposureProgram":2,"ISOSpeedRatings":400,"ExifVersion":"0221","DateTimeOriginal":"2016:11:26  17:16:57","DateTimeDigitized":"2016:11:26  17:16:57","ComponentsConfiguration":"\\u0001\\u0002\\u0003\\u0000","E'...;  public $published = '1'; public $filename =  'as-sam-9288_2_20230518_1691757047.jpg'; public $hits = NULL; public  $downloads = NULL; public $imgvotes = NULL; public $imgvotesum = NULL;  public $approved = NULL; public $useruploaded = NULL; public $access =  '1'; public $hidden = NULL; public $featured = NULL; public $ordering =  42; public $params =  '{"jg_dynamic_watermark":"-1","jg_showexifdata":"-1","jg_showgeotagging":"-1","jg_showiptcdata":"-1"}';  public $language = '*'; public $created_time = '2023-05-18 20:57:52';  public $created_by = '290'; public $modified_time = '2023-05-18  20:57:52'; public $modified_by = '290'; public $checked_out = NULL;  public $checked_out_time = NULL; public $metadesc = ''; public $metakey =  ''; public $robots = ''; public $tags = []; public $asset_id = 497 } ) | ...\ImageModel.php:505
14 | 0.3763 | 4163184 | Joomgallery\Component\Joomgallery\Administrator\Service\FileManager\FileManager->createImages( $source = 'C:\\wamp64\\www\\Joomla\\tmp\\AS-SAM_9288.JPG', $filename = 'as-sam-9288_2_20230518_1691757047.jpg', $cat = '7' ) | ...\Uploader.php:423
15 | 1.8511 | 4672784 | Joomgallery\Component\Joomgallery\Administrator\Service\IMGtools\IMtools->stream( $quality = 100, $base64 = FALSE, $html = ???, $type = ??? ) | ...\FileManager.php:258
16 | 1.8511 | 4672848 | Joomgallery\Component\Joomgallery\Administrator\Service\IMGtools\IMtools->write( $file = 'C:\\wamp64\\www\\Joomla/tmp/tmp_img.jpg', $quality = 100 ) | ...\IMtools.php:435
17 | 2.3304 | 4674344 | Joomgallery\Component\Joomgallery\Administrator\Service\IMGtools\IMGtools->checkValidImage( $img = 'C:\\wamp64\\www\\Joomla\\tmp\\tmp_img.jpg' ) | ...\IMtools.php:323
18 | 2.3306 | 4674288 | getimagesize( $filename = 'C:\\wamp64\\www\\Joomla\\tmp\\tmp_img.jpg' ) | ...\IMGtools.php:495

</font><!--EndFragment-->
</body>
</html>( ! ) Notice: getimagesize(): Error reading from C:\wamp64\www\Joomla\tmp\tmp_img.jpg! in C:\wamp64\www\Joomla\administrator\components\com_joomgallery\src\Service\IMGtools\IMGtools.php on line 495
Call Stack
#	Time	Memory	Function	Location
1	0.0001	368184	{main}( )	...\index.php:0
2	0.0002	368952	require_once( 'C:\wamp64\www\Joomla\administrator\includes\app.php )	...\index.php:32
3	0.0083	1386656	Joomla\CMS\Application\CMSApplication->execute( )	...\app.php:61
4	0.0086	1507872	Joomla\CMS\Application\AdministratorApplication->doExecute( )	...\CMSApplication.php:293
5	0.0477	2596784	Joomla\CMS\Application\AdministratorApplication->dispatch( $component = ??? )	...\AdministratorApplication.php:186
6	0.0512	2870760	Joomla\CMS\Component\ComponentHelper::renderComponent( $option = 'com_joomgallery', $params = ??? )	...\AdministratorApplication.php:143
7	0.0527	3075032	Joomla\CMS\Dispatcher\ComponentDispatcher->dispatch( )	...\ComponentHelper.php:361
8	0.0530	3096696	Joomgallery\Component\Joomgallery\Administrator\Controller\JoomFormController->execute( $task = 'ajaxsave' )	...\ComponentDispatcher.php:143
9	0.0530	3096760	Joomla\CMS\MVC\Controller\BaseController->execute( $task = 'ajaxsave' )	...\JoomFormController.php:87
10	0.0530	3096760	Joomgallery\Component\Joomgallery\Administrator\Controller\ImageController->ajaxsave( )	...\BaseController.php:678
11	0.0530	3096760	Joomla\CMS\MVC\Controller\FormController->save( $key = ???, $urlVar = ??? )	...\ImageController.php:107
12	0.0669	3516520	Joomgallery\Component\Joomgallery\Administrator\Model\ImageModel->save( $data = ['id' => '0', 'uploader' => 'tus', 'multiple' => '1', 'uuid' => '5d140b029646c57d5a683cbb6bda9b06', 'imgtitle' => 'title', 'filecounter' => '2', 'catid' => '7', 'published' => '1', 'imgauthor' => '', 'language' => '*', 'debug' => '0', 'imgtext' => '', 'access' => '1', 'tags' => []] )	...\FormController.php:675
13	0.3724	4159968	Joomgallery\Component\Joomgallery\Administrator\Service\Uploader\Uploader->createImage( $data_row = class Joomgallery\Component\Joomgallery\Administrator\Table\ImageTable { protected $_tbl = '#__joomgallery'; protected $_tbl_key = 'id'; protected $_tbl_keys = [0 => 'id']; protected $_db = class Joomla\Database\Mysql\MysqlDriver { public $name = 'mysql'; protected $nameQuote = '`'; protected $nullDate = '0000-00-00 00:00:00'; protected $utf8mb4 = TRUE; protected $mariadb = FALSE; protected $connection = class PDO { ... }; private ${Joomla\Database\DatabaseDriver}database = 'JoomlaTest'; public $serverType = 'mysql'; protected $count = 88; protected $cursor = NULL; protected $executed = FALSE; protected $limit = 0; protected $offset = 0; protected $options = [...]; protected $sql = class Joomla\Database\Mysql\MysqlQuery { ... }; protected $statement = class Joomla\Database\Pdo\PdoStatement { ... }; protected $tablePrefix = 'aji0d_'; protected $utf = TRUE; protected $errorNum = 0; protected $errorMsg = NULL; protected $transactionDepth = 0; protected $factory = class Joomla\Database\DatabaseFactory { ... }; protected $monitor = NULL; private ${Joomla\Database\DatabaseDriver}dispatcher = class Joomla\Event\Dispatcher { ... }; public $charset = 'utf8mb4' }; protected $_trackAssets = TRUE; protected $_rules = NULL; protected $_locked = FALSE; protected $_autoincrement = TRUE; protected $_columnAlias = ['published' => 'published']; protected $_jsonEncode = []; protected $_supportNullValue = FALSE; public $typeAlias = 'com_joomgallery.image'; protected $_errors = []; private ${Joomla\CMS\Table\Table}dispatcher = class Joomla\Event\Dispatcher { protected $events = [...]; protected $listeners = [...] }; public $form = class Joomla\CMS\Form\Form { protected $data = class Joomla\Registry\Registry { ... }; protected $errors = [...]; protected $name = 'image'; protected $options = [...]; protected $xml = class SimpleXMLElement { ... }; public $repeat = FALSE; private $databaseAwareTraitDatabase = NULL }; public $id = '164'; public $catid = '7'; public $alias = 'as-sam-9288-jpg-2'; public $imgtitle = 'AS-SAM_9288.JPG'; public $imgtext = ''; public $imgauthor = ''; public $imgdate = '2016:11:26 17:16:57'; public $imgmetadata = '{"exif":{"IFD0":{"ImageDescription":"SAMSUNG CSC","Make":"SAMSUNG","Model":"NX300","Orientation":1,"XResolution":"350\\/1","YResolution":"350\\/1","ResolutionUnit":2,"Software":"1.45 ","DateTime":"2016:11:26 17:16:57","YCbCrPositioning":2,"Exif_IFD_Pointer":254},"EXIF":{"ExposureTime":"1\\/40","FNumber":"35\\/10","ExposureProgram":2,"ISOSpeedRatings":400,"ExifVersion":"0221","DateTimeOriginal":"2016:11:26 17:16:57","DateTimeDigitized":"2016:11:26 17:16:57","ComponentsConfiguration":"\\u0001\\u0002\\u0003\\u0000","E'...; public $published = '1'; public $filename = 'as-sam-9288_2_20230518_1691757047.jpg'; public $hits = NULL; public $downloads = NULL; public $imgvotes = NULL; public $imgvotesum = NULL; public $approved = NULL; public $useruploaded = NULL; public $access = '1'; public $hidden = NULL; public $featured = NULL; public $ordering = 42; public $params = '{"jg_dynamic_watermark":"-1","jg_showexifdata":"-1","jg_showgeotagging":"-1","jg_showiptcdata":"-1"}'; public $language = '*'; public $created_time = '2023-05-18 20:57:52'; public $created_by = '290'; public $modified_time = '2023-05-18 20:57:52'; public $modified_by = '290'; public $checked_out = NULL; public $checked_out_time = NULL; public $metadesc = ''; public $metakey = ''; public $robots = ''; public $tags = []; public $asset_id = 497 } )	...\ImageModel.php:505
14	0.3763	4163184	Joomgallery\Component\Joomgallery\Administrator\Service\FileManager\FileManager->createImages( $source = 'C:\\wamp64\\www\\Joomla\\tmp\\AS-SAM_9288.JPG', $filename = 'as-sam-9288_2_20230518_1691757047.jpg', $cat = '7' )	...\Uploader.php:423
15	1.8511	4672784	Joomgallery\Component\Joomgallery\Administrator\Service\IMGtools\IMtools->stream( $quality = 100, $base64 = FALSE, $html = ???, $type = ??? )	...\FileManager.php:258
16	1.8511	4672848	Joomgallery\Component\Joomgallery\Administrator\Service\IMGtools\IMtools->write( $file = 'C:\\wamp64\\www\\Joomla/tmp/tmp_img.jpg', $quality = 100 )	...\IMtools.php:435
17	2.3304	4674344	Joomgallery\Component\Joomgallery\Administrator\Service\IMGtools\IMGtools->checkValidImage( $img = 'C:\\wamp64\\www\\Joomla\\tmp\\tmp_img.jpg' )	...\IMtools.php:323
18	2.3306	4674288	[getimagesize](http://www.php.net/function.getimagesize)( $filename = 'C:\\wamp64\\www\\Joomla\\tmp\\tmp_img.jpg' )	...\IMGtools.php:495



Save failed with the following error: Filename: AS-SAM_9288.JPG
Upload complete (File-Size: 7098.99 KB)

Imagetype: thumbnail
Used image processor: ImageMagick 7.1.0-51 Q16-HDRI x64 aea87b5:20221016 https://imagemagick.org
Resizing to height...
Auto-Orient image ...
Image manipulation successful.
Image-File (as-sam-9288_2_20230518_1691757047.jpg) of type 'thumbnail' successfully created.

Imagetype: detail
Used image processor: ImageMagick 7.1.0-51 Q16-HDRI x64 aea87b5:20221016 https://imagemagick.org
Resizing to height...
Auto-Orient image ...
Image manipulation successful.
Image-File (as-sam-9288_2_20230518_1691757047.jpg) of type 'detail' successfully created.

Imagetype: original
Used image processor: ImageMagick 7.1.0-51 Q16-HDRI x64 aea87b5:20221016 https://imagemagick.org
ERROR: Server problem. The following statement could not be executed in the php code: 'exec(magick convert -quality "100" "C:\wamp64\www\Joomla\tmp\AS-SAM_9288.JPG" "C:\wamp64\www\Joomla\tmp\tmp_img.jpg");'
Errors detected in processed image. Abort further processing.
Image-File (as-sam-9288_2_20230518_1691757047.jpg) of type 'original' could not be created.
There were some errors in the filesystem. Please check the image folders manually and resolve the problems.
Image-File (as-sam-9288_2_20230518_1691757047.jpg) of type 'thumbnail' successfully deleted.
Image-File (as-sam-9288_2_20230518_1691757047.jpg) of type 'detail' successfully deleted.
Image-File (as-sam-9288_2_20230518_1691757047.jpg) of type 'original' successfully deleted.

AlexanderSupp commented:
Now I have made 3 attempts with the same 40 images each.
At Nmb parallel processes 1, everything OK.
At Nmb parallel processes 5, 18 images had errors.
At Nmb parallel processes 10, 16 images had errors.
It wasn't always the same images that were faulty. I can't see any logic at this point. My processor has 12 cores and the allocation of activities will probably be crucial here.

AlexanderSupp commented:
> > When multi-uploading, there is a field at the bottom right: "Description" when I write something in there, I couldn't find it later in the individual images. What's the point of the field?
> 
> Confirmed. There was a bug in the code such that the description was not submitted with the form. Therefore no description text was saved. Should be solved now.

Is this solved?

AlexanderSupp commented:
Is this solved? Yes, but the individual description with the Pen-Symbol works not for me.

AlexanderSupp commented:
Is there a limit for uploads?
![Screenshot 2023-05-19 001551](https://github.com/JoomGalleryfriends/JG4-dev/assets/110494397/e53c92ca-03b4-4adb-a938-b2a2de7174f5)

eumel1602 commented:
> > If I don't assign a category before uploading, I get the error message: "Please fill out the required fields." But you don't see "red" which field is missing...
> 
> There is no nice way to fix that. I tried something. Is that good enough?

Yes it's OK !

--------
>  Is this solved? Yes, but the individual description with the Pen-Symbol works not for me.

Das Probmlem sollte gefixt sein. Ich habe nochmal alles versucht. Jede Beschreibung wird übernommen, ob das die Hauptbeschreibung unten rechts ist oder die einzelnen Beschreibungen mit dem Stift Symbol. Es funktioniert. 
Kannst du genauer Aussagen machen zum Browser, oder deinen Vorgehen speziell zu diesem Problem machen? Du musst jede Beschreibung extra speichern wenn du das Stift symbole benutzt!?

----
>  Is there a limit for uploads?

I tried to upload very small .jpg images. My limit is 64 pieces of images. After that, the upload no longer works.
There is no error message, only the note: Upload failed
?
0 of 0 files uploaded

eumel1602 commented:
Up to a certain size, the upload window does not increase although there is still space in this area. You can't make it bigger manually. bug? or exactly what you want? Is that specified and is it firmly written? Or would you rather not waste Manuel's time on it?'
(all Browser)
![breite](https://github.com/JoomGalleryfriends/JG4-dev/assets/96334985/a647a536-6ac8-4ec5-b2ec-69d0c4aa8748)

Elfangor93 commented:
>I uploaded 40 files with Nmb parallel processes 1 and everything works.
I uploaded the same 40 files with Nmb parallel processes 10, and I get errors.

Can you please give me the output of the browser console too?
Maybe there is a JavaScript error...

Elfangor93 commented:
>There is no error message, only the note:
"Upload failed ?
0 of 0 files uploaded"

Can you please add the output of the browser console?
I am pretty sure that there are some JavaScript errors...

Elfangor93 commented:
>You can't make it bigger manually. bug?

The size of the uppy dashboard is dynamically calculeated in the uppy app. If you want that to be changed you have to open an issue in the uppy repository (https://github.com/transloadit/uppy). I can not change that behavior in JoomGallery.

MrMusic commented:
First: 
Most of the problems should be fixed - thank you @Elfangor93 

Still a few problems for me:

1: When klicking on upload, there is a '404 Not Found' error in the browser console for each image. Example:
`HEAD http://localhost/jg4-dev/administrator/index.php?option=com_joomgallery&task=images.tusupload&uuid=5281b736a4c9d30c8f9ed27080412aec`
Nevertheless, all images are created.

2: Problem when uploading very large images (6000 x 4000px; about 10MB)
When using **GD** (surely there is an overflow of the memory limit): 
Javascript Error: `Uncaught (in promise) SyntaxError: JSON.parse: unterminated string at line 1 column 12 of the JSON data`
in webpack://uppy-uploader/src/index.js: 87:23

When using **ImageMagick**:
Also no upload possible.
Debug Information:
```
Upload of file xxx using Uppy failed.
This file is too large to upload. You can change the limits for your site in the component options.
Save failed with the following error: Unable to upload file.
```
Where it is a limit here?
When using Imagemagick should the php memory limit not matter?

3: In general, it should be noted that a image record is always created in the database, even if an error occurs during the upload and no images are created.

AlexanderSupp commented:
@eumel1602 

    Is this solved? Yes, but the individual description with the Pen-Symbol works not for me.

The problem should be fixed. I tried everything again. Each description will be adopted, whether that is the main description at the bottom right or the individual descriptions with the pencil icon. It works.
Can you make more precise statements about the browser, or make your approach specific to this problem? You have to save each description separately when using the pencil icon!?

Take a look at my Config. Simply, I select a file. I select the pencil-icon. I typed in Title, Description, Author. I select save. I select upload.
Title and Description will not place in the fields. Author is OK.

![Screenshot 2023-05-19 120518](https://github.com/JoomGalleryfriends/JG4-dev/assets/110494397/ae84f2b3-c7e2-4a7b-8c52-22eeb1387e84)
![Screenshot 2023-05-19 120246](https://github.com/JoomGalleryfriends/JG4-dev/assets/110494397/36612139-5c26-4d9c-ae51-776663618731)
![Screenshot 2023-05-19 115759](https://github.com/JoomGalleryfriends/JG4-dev/assets/110494397/a197d4e4-b624-4453-8ed0-78a7b8e345bc)
![Screenshot 2023-05-19 120021](https://github.com/JoomGalleryfriends/JG4-dev/assets/110494397/99bd79dd-ecbc-4da6-b44b-e07ea2289187)
![Screenshot 2023-05-19 120049](https://github.com/JoomGalleryfriends/JG4-dev/assets/110494397/81e80e7d-da29-4894-9b31-9b69cfaa6816)

AlexanderSupp commented:
> > I uploaded 40 files with Nmb parallel processes 1 and everything works.
> > I uploaded the same 40 files with Nmb parallel processes 10, and I get errors.
> 
> @AlexanderSupp Can you please give me the output of the browser console too? Maybe there is a JavaScript error...

I set Nmb parallel processes to 2. And I limited the images to 11.
So I hope you can find something that helps. It was my first time to use the browser console.

[console-export-2023-5-19_12-46-5.txt](https://github.com/JoomGalleryfriends/JG4-dev/files/11516143/console-export-2023-5-19_12-46-5.txt)


![Screenshot 2023-05-19 123921](https://github.com/JoomGalleryfriends/JG4-dev/assets/110494397/dde4def6-3010-4a03-bc37-2d20d88fb01c)
![Screenshot 2023-05-19 124648](https://github.com/JoomGalleryfriends/JG4-dev/assets/110494397/a4a19331-9731-40ca-a0d2-7c81d0310d13)

eumel1602 commented:
> AlexanderSupp: "Take a look at my Config. Simply, I select a file. I select the pencil-icon. I typed in Title, Description, Author. I select save. I select upload.
Title and Description will not place in the fields. Author is OK."

Did you notice that this is a bug of the upload method: 
>  Confirmed. Since this is an issue of the uppy library this has to be fixed in the uppy project. Bug is already reported in the 
> corresponding repository:
> https://github.com/transloadit/uppy/issues/4427

Elfangor93 commented:
>When klicking on upload, there is a '404 Not Found' error in the browser console for each image. Example:
HEAD http://localhost/jg4-dev/administrator/index.php?option=com_joomgallery&task=images.tusupload&uuid=5281b736a4c9d30c8f9ed27080412aec
Nevertheless, all images are created.

Yes. This is correct like that. If you upload the same image with uppy multiple times, uppy will send out a request to the TUS server checking if the image is already available in the destination location on the server. If not, the TUS server response back with a 404, telling the client that this image is not yet uploaded. 
For more information see https://tus.io/faq#how-does-tus-work

Elfangor93 commented:
>Debug Information:
Upload of file xxx using Uppy failed.
This file is too large to upload. You can change the limits for your site in the component options.
Save failed with the following error: Unable to upload file.
Where it is a limit here?

@MrMusic This limit comes from the com_media settings. We are using the filesystem-plugins to handle the filesystem. Since this plugins seem to read out some settings from com_media this limits also affect us now...
![grafik](https://github.com/JoomGalleryfriends/JG4-dev/assets/39154009/16769650-9f24-4bc0-9d5c-38abe3c61923)

AlexanderSupp commented:
> > Debug Information:
> > Upload of file xxx using Uppy failed.
> > This file is too large to upload. You can change the limits for your site in the component options.
> > Save failed with the following error: Unable to upload file.
> > Where it is a limit here?
> 
> @MrMusic This limit comes from the com_media settings. We are using the filesystem-plugins to handle the filesystem. Since this plugins seem to read out some settings from com_media this limits also affect us now... 
Information:
In all my test systems, I set Media Maximum Size (in MB) to zero.

AlexanderSupp commented:
> > AlexanderSupp: "Take a look at my Config. Simply, I select a file. I select the pencil-icon. I typed in Title, Description, Author. I select save. I select upload.
> > Title and Description will not place in the fields. Author is OK."
> 
> Did you notice that this is a bug of the upload method:
> 
> > Confirmed. Since this is an issue of the uppy library this has to be fixed in the uppy project. Bug is already reported in the
> > corresponding repository:
> > [transloadit/uppy#4427](https://github.com/transloadit/uppy/issues/4427)
> 
> Only one of the three things is saved at a time. (The joomgallery V4 can't do anything about that)

Thanks for this info. I haven't realized that.

Elfangor93 commented:
>In general, it should be noted that a image record is always created in the database, even if an error occurs during the upload and no images are created.

Yes. This is also true for single uploads. I have to think about a logic to rollback also the db creation if there is an error during image creation. But I would like to do that in a separate PR.
@MrMusic Can you please open a separate Issue for that?

Elfangor93 commented:
> 2: Problem when uploading very large images (6000 x 4000px; about 10MB)
> When using **GD** (surely there is an overflow of the memory limit):
> Javascript Error: `Uncaught (in promise) SyntaxError: JSON.parse: unterminated string at line 1 column 12 of the JSON data`
> in webpack://uppy-uploader/src/index.js: 87:23

No a memory limit overflow looks like this:
![grafik](https://github.com/JoomGalleryfriends/JG4-dev/assets/39154009/d7061778-a7c8-4e3c-a17f-0ae749971210)
What you get is an invalid json reply from the request. Can you please send me the content of the server request when this error happens? In there we should see what went wrong...

eumel1602 commented:
> > There is no error message, only the note:
> > "Upload failed ?
> > 0 of 0 files uploaded"
> 
> @eumel1602 Can you please add the output of the browser console? I am pretty sure that there are some JavaScript errors...

I hope you mean this:
![browserconsole](https://github.com/JoomGalleryfriends/JG4-dev/assets/96334985/2d7e6f8c-6c40-44fa-a45e-9e39b25a8f3f)

Elfangor93 commented:
@eumel1602 Just to clarify: You get the above posted JavaScript error when you add more than 64 images to the uppy dashboard and press the upload button?

eumel1602 commented:
> @eumel1602 Just to clarify: You get the above posted JavaScript error when you add more than 64 images to the uppy dashboard and press the upload button?

yes, exactly ! (in this case 99 pieces, small .jpg pictures...

Elfangor93 commented:
> You get the above posted JavaScript error when you add more than 64 images to the uppy dashboard and press the upload button?

Oke, I think this JavaScript errors are not solvable in an easy way. I have to shift this to a new PullRequest. I need to digg into uppy-plugin development for this. The error accures because of the hacky way I am doing the postprocessing of the uploads.

MrMusic commented:
I can confirm the problem when Nmb parallel processes is greater 1 and image procesor is ImageMagick. 
It is probably because it is trying to access the tmp/tmp_img.jpg file?

>This limit comes from the com_media settings...

Thank you for the clarification. I had not expected this.

reilldesign commented:
Image selection will be cleared if the form is invalid (miising title).

It would be great if the error is displayed but the image selection remains and does not need to be restarted.


![Image-Selection](https://github.com/JoomGalleryfriends/JG4-dev/assets/43771555/7b1c986c-713e-48e1-ba36-cc354a43dcbd)
